gdb-intro.md: first draft.

JohnsonLag opened the pull request:
Added link to the file in intro-to-c/index.md.

JohnsonLag committed:
gdb-intro.md: first draft.

Added link to the file in intro-to-c/index.md.

jerrettl commented on src/intro-to-c/debugging/gdb-intro.md:
```diff
@@ -0,0 +1,369 @@
+title: Introduction to GDB
+author: Johnson Laguerre
+---
+
```
Just an extra tidbit, maybe add what "GDB" stands for? (GNU debugger)

jerrettl commented on src/intro-to-c/debugging/gdb-intro.md:
````diff
@@ -0,0 +1,369 @@
+```
+
+No, not a segfault! This can't be possible--my code is *perfect*! Fine, I guess I'll just have to debug it. Time to start typing up print statements... or maybe I can try that debugger I've been reading about. What is it called again? Oh, yeah, GDB.
+
````
Love the intro to this point. Engaging yet informative and sets the scene.

jerrettl commented on src/intro-to-c/debugging/gdb-intro.md:
````diff
@@ -0,0 +1,369 @@
+Alright, first I have to recompile my code with the `-g` flag.
+
+```
+my-pc:~intro-to-c/assignments/pointers$ gcc -g my_file.c
````
Next bonus tidbit: using `-g` will add debugging symbols, but using `-ggdb` will add alllll the debugging symbols *specifically with GDB in mind.*

From the documentation:
```
       -ggdb
           Produce debugging information for use by GDB.  This means to use the most expressive format available
           (DWARF, stabs, or the native format if neither of those are supported), including GDB extensions if at all
           possible.
```

jerrettl commented on src/intro-to-c/debugging/gdb-intro.md:
````diff
@@ -0,0 +1,369 @@
+Reading symbols from a.out...
+(gdb)
+```
+
````
I might consider adding a little blurb here that sets the scene for GDB. Something like it acts like its own mini command-line environment with its own set of commands. A lot of people learning to use GDB for the first time may not be super comfortable with command line environments still.

jerrettl reviewed (commented):
Awesome stuff! This is one of those guides that I've been thinking about for a long time. My only thoughts to this are to be a little more explicit with the exact commands used, as this will be a completely foreign experience for some people. It is mainly self-explanatory for maybe people like you or I, but I can see this getting people lost if they've never seen this before.

So just little explanations about what you are about to do for each step: i.e., "The `run` command will run your program as if you were running it normally in the terminal."

And then when you observe the output of GDB, I would point out what leads you to the conclusions you have. Like on line 160: "Since `i` is 0, this means the segmentation fault occurred during my first time through the loop..."

Again, I can understand myself from reading, but people new to debugging may appreciate the little extra verbosity. Just my 2¢.

Something else you may want to consider (purely aesthetic) is the `.console` class for command-line output. See [here](https://github.com/wikiknights/cs-materials/blob/0853922bf9a1321157c6b7127a9dbada29204b9c/doc/sample-file.md?plain=1#L323) for details if interested.

I've also thrown in just a few little tidbits that you might want to add, but that is completely up to you.

Nice work Johnson!

JohnsonLag committed:
gdb-intro.md: second draft.

JohnsonLag commented on src/intro-to-c/debugging/gdb-intro.md:
```diff
@@ -185,9 +199,17 @@ Breakpoint 1, main () at my_file.c:6
 
+> The `break` command lets you set breakpoints, specific points where you want your program to stop while it's running. A breakpoint gives you a location from which you can choose how and where you want to proceed, whether line-by-line or multiples lines at a time.
+
+> Here, we pass `break` a function name as an argument, but you can also pass it a line number (e.g., `break 6`). If you wanted to be more specific, you could pass it a location written as a file name, colon, and either a function name or a line number (e.g., `break my_file.c:main`, `break my_file.c:6`).
```
Do you think I should I keep this sentence about being "more specific"? It's more useful to know when you're compiling multiple source files together, but I think you typically compile single files in Intro to C.

jerrettl replied:
I think this is good extra information. No harm in keeping it, since this lesson is placed towards the end of the lesson list. A student may be jumping into CS1 at this point.

JohnsonLag commented:
Thank you!

I wasn't sure if too many details would be overwhelming/distracting, but that makes sense. I used blockquotes to provide "explanation segments" underneath the code blocks.

There are certain parts of the guide I'd like your opinion on, I'm working on making comments for those parts.

JohnsonLag commented on src/intro-to-c/debugging/gdb-intro.md:
````diff
@@ -157,9 +165,13 @@ $3 = (int *) 0x4
-```
+Okay, since `i` is 0, that means the segfault happened on my first time through the loop. I can also see that `current` *clearly* isn't pointing to `arr`'s address.
+
+In fact, the address in `current` is only 4 bytes away from `NULL`, so it looks like I dereferenced a `NULL` pointer then did the post-increment. I'll continue and see if anything else pops up.
````
This explanation about `current` is actually wrong, so I need to edit it. (I thought the dereferencing happened before the post-increment, but it's actually the other way around.)

I also plan on updating the example code so that the dereferencing and post-incrementing happen on separate lines. One reason is so when I print `current`, we'll be able to clearly see that it was pointing to 0x0 when I dereferenced it. Another reason is that based on what I see about operator precedence for dereferencing and post-incrementing, the example code shouldn't be working in the end, even though it is.

---

Also, I noticed that the [Operators file](https://github.com/wikiknights/cs-materials/blob/main/src/intro-to-c/operators/operators-intro.md#arithmetic-operators) mentions the increment (`++`) and decrement (`--`) operators but doesn't explicitly mention pre- and post-incrementing. Is that something you think we should add?

jerrettl replied:
I see what you mean, that is something I am sure you can take on, so I won't say anything about the lines on `current`.

For the operators, that is a good point, and I would add in something about pre- and post-incrementing if you are able.

jerrettl commented on src/intro-to-c/debugging/gdb-intro.md:
````diff
@@ -243,9 +267,11 @@ multiply_by_two (current=0x0, arr=0x7fffffffdfe0) at my_file.c:37
 (gdb)
 ```
 
+> The `until` command lets you continue running your program until you reach a certain line. The program will still stop on any breakpoints.
````
I see no explanation on the `step` command. Kind of self-explanatory, but might want to add that in.

jerrettl reviewed (commented):
I think these new additions add a lot of clarity to this page. I can definitely see myself referring to this for students wrapping up Intro to C to use to learn how to debug. Awesome work on this!

Something I was thinking about is that since this covers a program that uses pointers and arrays, students will need to understand those concepts before this page will make any sense. Not that this is necessarily an issue, because I think a more broad "how to debug" page is in order that uses simple print-based debugging. (Not really related to this page specifically, but it was something I wanted to throw out there.)

JohnsonLag committed:
gdb-intro.md: third draft.

Changed explanation for segmentation fault. Added tip for repeating commands. Added explanation for the `step` command.

JohnsonLag commented:
Thank you!

I see what you mean. When I made the page, I was thinking about what might be a good way to show that GDB is useful, so I went with pointers since I thought that would bring in the kind of "complexity" that would push someone to want to use a debugger (e.g., seeing a segmentation fault and wanting to see where exactly things went wrong).

If nobody else is planning to, I was thinking about possibly doing the "General debugging" page after this. Did you have specific scenarios or examples you think would be good to cover in it?

For the Operators page, I'll see if I can update it sometime this week.

jerrettl commented:
> When I made the page, I was thinking about what might be a good way to show that GDB is useful, so I went with pointers since I thought that would bring in the kind of "complexity" that would push someone to want to use a debugger (e.g., seeing a segmentation fault and wanting to see where exactly things went wrong).

Makes complete sense to me!

> Did you have specific scenarios or examples you think would be good to cover in it?

There were a couple of ideas I was thinking for this, they are all relatively simple situations with a little spot or two that can throw things off.

One example - Can't figure out why my program isn't printing anything:

This is supposed to calculate change.

``` c
int main(void)
{
    double change;
    int dollars, quarters, dimes, nickels, pennies;

    printf("Enter the amount of change: ");
    scanf("%lf", &change);

    while (change != 0)
    {
        if (change >= 1)
        {
            dollars++;
            change -= 1;
        }
        else if (change >= 0.25)
        {
            quarters++;
            change -= 0.25;
        }
        else if (change >= 0.10)
        {
            dimes++;
            change -= 0.10;
        }
        else if (change >= 0.05)
        {
            nickels++;
            change -= 0.05;
        }
        else if (change >= 0.01)
        {
            pennies++;
            change -= 0.01;
        }
    }

    printf("%d Dollars\n", dollars);
    printf("%d Quarters\n", quarters);
    printf("%d Dimes\n", dimes);
    printf("%d Nickels\n", nickels);
    printf("%d Pennies\n", pennies);

    return 0;
}
```

Another example - Debugging an off-by-one error that can cause stack smashing or unexpected behavior:

``` c
int main(void)
{
    int array1[10];
    int array2[10];
    int n = 10;

    // Initialize array1 with 5 and array2 with 10
    for (int i = 1; i <= n; i++)
    {
        array2[i] = 10;
        array1[i] = 5;
    }

    // Print all the array values
    for (int i = 1; i <= n; i++)
    {
        printf("array1[%d] = %d\n", i, array1[i]);
    }
    for (int i = 1; i <= n; i++)
    {
        printf("array2[%d] = %d\n", i, array1[i]);
    }

    return 0;
}
```

Another example ("Why isn't this program working right?"):

There are actually two problems with this one.

``` c
int main(void)
{
    int i;
    char name[100] = "banana";

    // This should print the name as a grid with 5 rows like:
    // b a n a n a
    // b a n a n a
    // b a n a n a
    // b a n a n a
    // b a n a n a

    for (i = 0; i < 5; i++)
    {
        for (i = 0; i < 100; i++)
        {
            printf("%c ", name[i]);
        }
        printf("\n");
    }

    return 0;
}
```

JohnsonLag commented:
I like the ideas, I'll see if I can start working on the page soon.
